meteor/client/ui/ScriptView/ScriptView.tsx

```diff
@@ -0,0 +1,667 @@
+import { PeripheralDevicesAPI, callPeripheralDeviceFunction } from '../../lib/clientAPI'
+import { doModalDialog } from '../../lib/ModalDialog'
+import { SupportPopUp } from '../SupportPopUp'
+import { Prompt } from 'react-router-dom'
+import { t } from 'i18next'
+import { NotificationCenterPanel } from '../../lib/notifications/NotificationCenterPanel'
+import { AdLibPiece } from '../../../lib/collections/AdLibPieces'
+import { SegmentContextMenu } from '../SegmentTimeline/SegmentContextMenu'
+
+export interface LayerGroups<T, P> {
+	primaryGroup: T
+	overlayGroup: T
+	audioGroup: T
+	otherGroup: T
+	adlibGroup: P
+}
+
+export interface OutputGroups<T, P> {
+	[key: string]: LayerGroups<T, P>
+}
+
+interface ScriptViewProps {
+	match?: {
+		params: {
+			rundownId: RundownPlaylistId
+		}
+	}
+	rundownId?: RundownId
+	inActiveRundownView?: boolean
+	onlyShelf?: boolean
+}
+interface ScriptViewState {
+	timeScale: number
+	studioMode: boolean
+	bottomMargin: string
+	followLiveSegments: boolean
+	manualSetAsNext: boolean
+	subsReady: boolean
+	isNotificationsCenterOpen: boolean
+	isSupportPanelOpen: boolean
+	isInspectorShelfExpanded: boolean
+	isClipTrimmerOpen: boolean
+	selectedPiece: AdLibPieceUi | PieceUi | undefined
+	rundownLayout: RundownLayout | undefined
+	contextMenuContext: IContextMenuContext | null
+}
+
+interface ScriptViewTrackedProps {
+	rundownId: RundownId
+	rundown: Rundown | undefined
+	rundownPlaylistId?: RundownPlaylistId
+	playlist?: RundownPlaylist
+	segments: Segment[]
+	parts: Part[]
+	pieces: Piece[]
+	adlibs: AdLibPiece[]
+	studio?: Studio
+	showStyleBase?: ShowStyleBase
+	rundownLayouts?: Array<RundownLayoutBase>
+	buckets: Bucket[]
+	casparCGPlayoutDevices?: PeripheralDevice[]
+	rundownLayoutId?: RundownLayoutId
+	activeLayerGroups: OutputGroups<boolean, boolean>
+}
+
+export const ScriptView = translateWithTracker<ScriptViewProps, ScriptViewState, ScriptViewTrackedProps>(
+	(props: ScriptViewProps) => {
+		let rundownId
+		if (props.match && props.match.params.rundownId) {
+			rundownId = decodeURIComponent(unprotectString(props.match.params.rundownId))
+		} else if (props.rundownId) {
+			rundownId = props.rundownId
+		}
+
+		const rundown = Rundowns.findOne(rundownId)
+		let playlist: RundownPlaylist | undefined
+		let segments: Segment[] = []
+		let parts: Part[] = []
+		let pieces: Piece[] = []
+		let adlibs: AdLibPiece[] = []
+
+		let showStyleBase: ShowStyleBase | undefined
+		let activeLayerGroups: OutputGroups<boolean, boolean> = {}
+
+		if (rundown) {
+			showStyleBase = ShowStyleBases.findOne(rundown.showStyleBaseId)
+			playlist = RundownPlaylists.findOne(rundown.playlistId)
+		}
+
+		let studio: Studio | undefined
+		if (playlist) {
+			studio = Studios.findOne({ _id: playlist.studioId })
+			segments = playlist.getSegments({
+				isHidden: {
+					$ne: true,
+				},
+			})
+
+			if (segments.length > 0) {
+				parts = segments.map((seg) => seg.getParts()).reduce((acc, val) => acc.concat(val), [])
```

Comment:
Parts undergo a lot of changes during playback. It would be good to limit the amount of fields this tracker will be reactive to (perhaps at least excluding some of the `timings` fields) or the amount of changes will kill the browser during any sort of an update.